Expose library version, move it out of options

gsoltis opened the pull request:
Expose library version as a constant. Remove it from FIROptions, since it is not per-options instance, and it doesn't make sense to set it.

wilhuff commented on Firebase/Core/Public/FIRVersion.h:
```diff
@@ -0,0 +1,20 @@
+#import <Foundation/Foundation.h>
+
+// The version of the Firebase SDK.
+extern NSString *const kFIRLibraryVersionID;
```
I see your declaration of kFIRLibraryVersionID here but I don't see a definition. Are you missing a file?

gsoltis replied:
No, it turns out it did already exist. It's in FIROptions.m[0]. It is possible that we would like a new home for it though. I'm hoping someone from the core team can give some guidance there.

This constant is also not currently tied to the one actually used to build the sdk by anything other than policy. That being said, it does appear to be current, so that might be good enough for now.

0: https://github.com/firebase/firebase-ios-sdk/blob/9ad65fe00f78302aa025e3e9ff753fbc0989599f/Firebase/Core/FIROptions.m#L42

paulb777 commented:
Can you explain more about the reasoning for this change? Why do you care about the FirebaseCore version?

If anything, I would like to see FirebaseCore adopt the FirebaseFirestore versioning approach using build time compile flags that don't require source changes.

wilhuff commented:
The issue here is that we're supposed to be reporting the Firebase-wide version to the backend. We also want to include the Firebase version in our log lines. The Firestore version is largely an implementation detail that nobody really cares about. Logging the Firebase-wide version makes it possible to go back to release notes, see if you're up-to-date, etc.

gsoltis commented:
Indeed, I think an even better approach would be to make available entire-SDK version that is used to build the whole thing. But, I couldn't see a way to thread it through easily. At the same time, I noticed that this constant is supposed to be kept in sync, so I figured it might work for now.

It also seemed slightly misplaced on FIROptions and with a setter. Different options instances would never have different sdk versions, so I think it is better as just a simple constant. And obviously we would not ever want to programmatically set the version.

paulb777 commented:
OK, but the FIROptions version isn't the Firebase-wide version - It's the FirebaseCore version.

A Firebase-wide version would need to be implemented in Firebase.h since that is the only file that is part of overall Firebase.

I'd like to see us merge FirebaseCore into Firebase, but that is not currently the implementation.

gsoltis commented:
I see. In that case, is there a way to get the correct version into Firebase.h?

gsoltis commented:
Ok, please take another look. I've added both Firebase_VERSION and FIRCore_VERSION to the podspec, as well as FIRVersion.h. If this is accepted, we'll need to add them to the internal build as well.

I've reverted the changes to FIROptions and test, although it seems like those might be nice as an independent change?

This change also includes adding the version to FIRLogger output. I have no strong preferences on the format of the log line, so if anyone would prefer something different, I'm happy to change it.

Once this is accepted, I will update Firestore to send the proper version in the GRPC header.

paulb777 commented on Firebase/Core/Private/FIRVersion.m:
```diff
@@ -0,0 +1,39 @@
+        (const unsigned char *const)STR(Firebase_VERSION);
+
+const unsigned char *const FirebaseCoreVersionString =
+        (const unsigned char *const)STR(FIRCore_VERSION);
```
Add newline to end of file

gsoltis replied:
Done.

paulb777 commented on Firebase/Core/Private/FIRVersion.m:
```diff
@@ -0,0 +1,39 @@
+/*
```
Current convention is Private directory is only for headers. FIRVersion.m should be in Firebase/Core/

gsoltis replied:
Ah, got it. Done.

paulb777 commented on Firebase/Core/Public/FIRVersion.h:
```diff
@@ -0,0 +1,23 @@
+/*
```
Putting FIRVersion.h may require an API review.  @ryanwilson ?

gsoltis replied:
I moved it to Private. It probably would make sense to move it to Public/ at some point though.

paulb777 commented on FirebaseCore.podspec:
```diff
@@ -28,4 +28,7 @@ Firebase Core includes FIRApp and FIROptions which provide central configuration
   s.framework = 'SystemConfiguration'
   s.dependency 'GoogleToolboxForMac/NSData+zlib', '~> 2.1'
+  s.pod_target_xcconfig = {
+    'OTHER_CFLAGS' => '-DFIRCore_VERSION=' + s.version.to_s + ' -DFirebase_VERSION=0.0.1-dev'
```
Firebase version should be 4.8.0.  Part of the iCore release process is synchronizing the open source development build with the official last Firebase public release with the glue in Example/Podfile and Firestore/ExamplePodfile.

gsoltis replied:
Done.

paulb777 commented on Firebase/Core/Private/FIRVersion.m:
```diff
@@ -0,0 +1,39 @@
+const unsigned char *const FirebaseVersionString =
+        (const unsigned char *const)STR(Firebase_VERSION);
+
+const unsigned char *const FirebaseCoreVersionString =
```
Is the idea to modify the FIROptions FirebaseCore option handling in a separate PR?

paulb777 commented:
If we want to defer the API question, FIRVersion.h could be added to Private instead.

gsoltis commented:
Ok, I think I've managed to rebase on top of master.

ryanwilson commented on Firebase/Core/FIRVersion.m:
```diff
@@ -0,0 +1,39 @@
+
+
+#ifndef Firebase_VERSION
+#error "Firebase_VERSION is not defined: add -DFirebase_VERSION=... to the build invocation"
```
@paulb777 Will this cause trouble for future build systems that don't use CocoaPods?

paulb777 replied:
@ryanwilson This should be fine. We already require any build system to add -D versions for multiple Firebase components.  This ifndef is a great way to catch it at build time.

Separately, we do often use #ifdef COCOAPODS in a binary way that will have to be reviewed with a third build system.

ryanwilson replied:
Thanks!
